chain/client/src/sync/state/network.rs

```diff
@@ -0,0 +1,294 @@
+                .as_ref()
+                .is_some_and(|expecting_peer_id| expecting_peer_id == &peer_id)
+            {
+                let value = self.pending_requests.remove(&key).unwrap();
+                let _ = value.sender.send(data);
+                return Ok(());
+            }
+        }
+        Err(near_chain::Error::Other("Unexpected message".to_owned()))
+    }
+
+    /// Sets the peers that are eligible for querying state sync headers/parts.
+    pub fn set_highest_peers(&mut self, peers: Vec<PeerId>) {
+        self.highest_height_peers = peers;
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+enum PartIdOrHeader {
+    Part { part_id: u64 },
+    Header,
+}
+
+impl StateSyncDownloadSourcePeer {
+    async fn try_download(
+        clock: Clock,
+        request_sender: AsyncSender<PeerManagerMessageRequest, PeerManagerMessageResponse>,
+        key: PendingPeerRequestKey,
+        store: Store,
+        state: Arc<Mutex<StateSyncDownloadSourcePeerSharedState>>,
+        cancel: CancellationToken,
+        request_timeout: Duration,
+        handle: Arc<TaskHandle>,
+    ) -> Result<ShardStateSyncResponse, near_chain::Error> {
+        handle.set_status("Preparing request");
+
+        // Sender/receiver pair used to await for the peer's response.
+        let (sender, receiver) = oneshot::channel();
+
+        let network_request = {
+            let mut state_lock = state.lock().unwrap();
+            let (network_request, state_value) = match &key.kind {
+                PartIdOrHeader::Part { part_id } => {
+                    let prev_hash = *store
+                        .get_ser::<BlockHeader>(DBCol::BlockHeader, key.sync_hash.as_bytes())?
+                        .ok_or_else(|| {
+                            near_chain::Error::DBNotFoundErr(format!(
+                                "No block header {}",
+                                key.sync_hash
+                            ))
+                        })?
+                        .prev_hash();
+                    let prev_prev_hash = *store
+                        .get_ser::<BlockHeader>(DBCol::BlockHeader, prev_hash.as_bytes())?
+                        .ok_or_else(|| {
+                            near_chain::Error::DBNotFoundErr(format!(
+                                "No block header {}",
+                                prev_hash
+                            ))
+                        })?
+                        .prev_hash();
+                    let network_request = PeerManagerMessageRequest::NetworkRequests(
+                        NetworkRequests::StateRequestPart {
+                            shard_id: key.shard_id,
+                            sync_hash: key.sync_hash,
+                            sync_prev_prev_hash: prev_prev_hash,
+                            part_id: *part_id,
+                        },
+                    );
+                    let state_value = PendingPeerRequestValue { peer_id: None, sender };
```

Comment:
In the handling for NetworkRequests::StateRequestPart we select a specific peer from which to request the part, so it should be possible to store that and verify that the response comes back from the expected peer. However, it might be a bit ugly to pass the selected peer id back here from the network side of things, and I expect to redo how the state headers work soon, so I am fine with just leaving this as-is for now.

Reply:
Yeah.. this part is awkward indeed. I also don't know what to do about it right now.